build.cake

```diff
@@ -0,0 +1,567 @@
+
+// Working directory
+var workingDirectory = (new CakeEnvironment()).WorkingDirectory;
+
+// System specific shell configuration
+var shell = IsRunningOnWindows() ? "powershell" : "bash";
+var shellArgument = IsRunningOnWindows() ? "/Command" : "-C";
+var shellExtension = IsRunningOnWindows() ? "ps1" : "sh";
+
+/// <summary>
+///  Class representing build.json
+/// </summary>
+public class BuildPlan
+{
+    public IDictionary<string, string[]> TestProjects { get; set; }
+    public string BuildToolsFolder { get; set; }
+    public string ArtifactsFolder { get; set; }
+    public string DotNetFolder { get; set; }
+    public string DotNetInstallScriptURL { get; set; }
+    public string DotNetChannel { get; set; }
+    public string DotNetVersion { get; set; }
+    public string[] Frameworks { get; set; }
+    public string[] Rids { get; set; }
+    public string MainProject { get; set; }
+}
+var buildPlan = DeserializeJsonFromFile<BuildPlan>($"{workingDirectory}/build.json");
+
+// Folders and tools
+var dotnetFolder = $"{workingDirectory}/{buildPlan.DotNetFolder}";
+var dotnetcli = $"{dotnetFolder}/cli/dotnet";
+var toolsFolder = $"{workingDirectory}/tools";
+var xunitRunner = "xunit.runner.console";
+
+var sourceFolder = $"{workingDirectory}/src";
+var testFolder = $"{workingDirectory}/tests";
+
+var artifactFolder = $"{workingDirectory}/{buildPlan.ArtifactsFolder}";
+var publishFolder = $"{artifactFolder}/publish";
+var logFolder = $"{artifactFolder}/logs";
+var packageFolder = $"{artifactFolder}/package";
+
+/// <summary>
+///  Retrieve the default local RID from .NET CLI.
+/// </summary>
+/// <param name="dotnetcli">Full path to .NET CLI binary</param>
+/// <returns>Default RID</returns>
+string GetLocalRuntimeID(string dotnetcli)
+{
+    var process = StartAndReturnProcess(dotnetcli, 
+        new ProcessSettings
+        { 
+            Arguments = "--version",
+            RedirectStandardOutput = true
+        });
+    process.WaitForExit();
+    if (process.GetExitCode() != 0)
+    {
+        throw new Exception("Failed to get run dotnet --version");
+    }
+    foreach (var line in process.GetStandardOutput())
+    {
+        if (!line.Contains("Runtime Id"))
+        {
+            continue;
+        }
+        var colonIndex = line.IndexOf(':');
+        return line.Substring(colonIndex + 1).Trim();
+    }
+    throw new Exception("Failed to get default RID for system");
+}
+
+/// <summary>
+///  Match the local RID with the ones specified in build.json.
+///  Return exact match if found.
+///  Return first OS match (without version number) otherwise.
+///  Report error if no OS match found.
+/// </summary>
+/// <param name="dotnetcli">Full path to .NET CLI binary</param>
+/// <param name="buildPlan">BuildPlan from build.json</param>
+/// <returns>Matched RID</returns>
+string MatchLocalRuntimeID(string dotnetcli, BuildPlan buildPlan)
```

Comment:
If user do a local build, which means he only build for current platform. `dotnet restore` doesn't need a `--runtime` parameter. We can simply omit it. In this situation, the build process shouldn't build for any other platform, because it is user's choice to build for local platform.

Imagine it works like this:

```
$ .\build.sh   # user choose to build locally, then only one platform is build, test and published.
$ .\build.sh --all # user choose to build for all platforms in build.json RIDs list. This is more like a CI scenario
```